File: 0112_path_sum/solution.cpp
class Solution {
    int fs = 0;
public:
    bool hasPathSum(TreeNode* root, int sum) {
        if(root == NULL)
            return false;
        if(root->left != NULL){
            this->fs += root->val;
            if(this->hasPathSum(root->left, sum))
                return true;
            this->fs -= root->val;
        }
        if(root->right != NULL){
            this->fs += root->val;
            if(this->hasPathSum(root->right, sum))
                return true;
            this->fs -= root->val;
        }
        if(root->left == NULL && root->right == NULL){
            this->fs += root->val;
            if(this->fs == sum)
                return true;
            this->fs -= root->val;
        }
        return false;
    }
};

File: 0112_path_sum/solution.py
class Solution:
    def __init__(self):
        self.fs = 0
    def hasPathSum(self, r: TreeNode, s: int) -> bool:
        if r is None:
            return False
        if (r.left is not None):
            self.fs += r.val
            if(self.hasPathSum(r.left, s)):
                return True
            self.fs -= r.val
        if (r.right is not None):
            self.fs += r.val
            if(self.hasPathSum(r.right, s)):
                return True
            self.fs -= r.val
        if (r.left is None) and (r.right is None):
            self.fs += r.val
            if(self.fs == s):
                return True
            self.fs -= r.val          

